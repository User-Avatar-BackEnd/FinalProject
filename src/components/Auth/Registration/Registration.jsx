import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { Formik, Form } from 'formik';
import * as Yup from 'yup';
import AuthInput from '../AuthInput/AuthInput';
import ErrorMessage from '../../ErrorMessage/ErrorMessage';
import API from '../../../config/API';
import errorsDescription from '../../../config/APIErrorsDescription';

import styles from './Registration.module.scss';

const RegistrationSchema = Yup.object().shape({
  email: Yup.string().email('Invalid email').required('Required'),
  password: Yup.string()
    .strict(false)
    .trim()
    .min(5, 'Must be 5 characters or more')
    .max(64, 'Must be 64 characters or less')
    .matches(/^[a-zA-Z0-9_.-]+$/, 'Can only contain letters, numbers and underscores')
    .required('Required'),
  repeatPassword: Yup.string()
    .strict(false)
    .trim()
    .oneOf([Yup.ref('password'), null], 'Passwords must match')
    .required('Required'),
  username: Yup.string()
    .strict(false)
    .trim()
    .min(5, 'Must be 5 characters or more')
    .max(14, 'Must be 14 characters or less')
    .matches(/^[a-zA-Z0-9_.-]+$/, 'Can only contain letters, numbers and underscores')
});

const Registration = ({ onLogin }) => {
  const [serverError, setServerError] = useState('')

  const clearError = () => {
    setServerError('')
  }

  return (
    <div className={styles.Registration}>
      <h2>Sign up</h2>
      <Formik
        initialValues={{ email: '', password: '', repeatPassword: '', username: '' }}
        validationSchema={RegistrationSchema}
        onSubmit={(values, { setSubmitting }) => {
          const data = {
            email: values.email,
            password: values.password.trim()
          }
          if (values.username) {
            data.login = values.username
          }

          API.post('/auth/register', data)
            .then(response => {
              onLogin(response.data)
            })
            .catch(error => {
              setServerError(
                errorsDescription.auth[error.response.data] ?? errorsDescription.default
              )
            })
        }}
      >
        {({ isSubmitting, errors, touched }) => (
          <Form onChange={clearError}>
            <AuthInput type='email' id='email' name='email' label='Email' placeholder='Email' required={true} autoComplete='off' />
            <AuthInput type='password' id='password' name='password' label='Password' placeholder='Password' required={true} />
            <AuthInput type='password' id='repeatPassword' name='repeatPassword' label='Repeat password' placeholder='Repeat password' required={true} />
            <AuthInput type='text' id='username' name='username' label='Username (optional)' placeholder='Username (optional)' required={false} />
            <button type='submit' >
              Submit
            </button>
          </Form>
        )}
      </Formik>
      <div className={styles.loginButton}>
        <Link to={'/login'} className={styles.loginLink}>Already have an account? Sign in</Link>
      </div>
      <ErrorMessage text={serverError} type={'form'} />
    </div>
  );

}

export default Registration;