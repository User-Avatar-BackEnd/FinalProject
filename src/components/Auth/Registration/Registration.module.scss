.Registration {
  width: 70%;
  margin: 0 auto;

  & form {
    display: flex;
    flex-direction: column;

    & input {

    }

    & button[type=submit] {
      margin: 0 auto;
      color: #fff;
      width: 140px;
      height: 30px;
      background-color: dodgerblue;
      border: 1px solid #fff;
      border-radius: 5px;
      cursor: pointer;
      transition: background-color .3s;

      &:hover {
        background-color: darken(dodgerblue, 20%);
      }
    }
  }

  & .loginButton {
    text-align: center;

    & .loginLink {
      color: dodgerblue;
      font-size: 12px;
      text-decoration: underline;
      cursor: pointer;

      &:hover {
        color: blue;
      }
    }
  }
}