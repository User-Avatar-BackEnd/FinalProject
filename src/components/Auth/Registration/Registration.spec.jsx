import React from 'react';
import { shallow } from 'enzyme';
import { Formik } from 'formik';

import Registration from './Registration';

describe('Registration component', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Registration />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('should render Sign up form', () => {
    const formik = wrapper.find(Formik);

    expect(formik.length).toBe(1);

    expect(wrapper).toMatchSnapshot();
  });
});