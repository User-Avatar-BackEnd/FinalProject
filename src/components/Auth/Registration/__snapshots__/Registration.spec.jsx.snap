// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registration component should render Sign up form 1`] = `
<div
  className="Registration"
>
  <h2>
    Sign up
  </h2>
  <Formik
    initialValues={
      Object {
        "email": "",
        "password": "",
        "repeatPassword": "",
        "username": "",
      }
    }
    onSubmit={[Function]}
    validationSchema={
      ObjectSchema {
        "_blacklist": ReferenceSet {
          "list": Set {},
          "refs": Map {},
        },
        "_blacklistError": undefined,
        "_excludedEdges": Array [],
        "_nodes": Array [
          "username",
          "repeatPassword",
          "password",
          "email",
        ],
        "_sortErrors": [Function],
        "_typeError": [Function],
        "_whitelist": ReferenceSet {
          "list": Set {},
          "refs": Map {},
        },
        "_whitelistError": undefined,
        "conditions": Array [],
        "deps": Array [],
        "exclusiveTests": Object {},
        "fields": Object {
          "email": StringSchema {
            "_blacklist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_blacklistError": undefined,
            "_mutate": undefined,
            "_typeError": [Function],
            "_whitelist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_whitelistError": undefined,
            "conditions": Array [],
            "deps": Array [],
            "exclusiveTests": Object {
              "email": false,
              "required": true,
            },
            "spec": Object {
              "abortEarly": true,
              "nullable": false,
              "presence": "required",
              "recursive": true,
              "strict": false,
              "strip": false,
            },
            "tests": Array [
              [Function],
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
            "type": "string",
          },
          "password": StringSchema {
            "_blacklist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_blacklistError": undefined,
            "_mutate": undefined,
            "_typeError": [Function],
            "_whitelist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_whitelistError": undefined,
            "conditions": Array [],
            "deps": Array [],
            "exclusiveTests": Object {
              "matches": false,
              "max": true,
              "min": true,
              "required": true,
              "trim": false,
            },
            "spec": Object {
              "abortEarly": true,
              "nullable": false,
              "presence": "required",
              "recursive": true,
              "strict": false,
              "strip": false,
            },
            "tests": Array [
              [Function],
              [Function],
              [Function],
              [Function],
              [Function],
            ],
            "transforms": Array [
              [Function],
              [Function],
            ],
            "type": "string",
          },
          "repeatPassword": StringSchema {
            "_blacklist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_blacklistError": undefined,
            "_mutate": undefined,
            "_typeError": [Function],
            "_whitelist": ReferenceSet {
              "list": Set {
                null,
              },
              "refs": Map {
                "password" => Reference {
                  "getter": [Function],
                  "isContext": false,
                  "isSibling": true,
                  "isValue": false,
                  "key": "password",
                  "map": undefined,
                  "path": "password",
                },
              },
            },
            "_whitelistError": [Function],
            "conditions": Array [],
            "deps": Array [],
            "exclusiveTests": Object {
              "required": true,
              "trim": false,
            },
            "spec": Object {
              "abortEarly": true,
              "nullable": false,
              "presence": "required",
              "recursive": true,
              "strict": false,
              "strip": false,
            },
            "tests": Array [
              [Function],
              [Function],
            ],
            "transforms": Array [
              [Function],
              [Function],
            ],
            "type": "string",
          },
          "username": StringSchema {
            "_blacklist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_blacklistError": undefined,
            "_typeError": [Function],
            "_whitelist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_whitelistError": undefined,
            "conditions": Array [],
            "deps": Array [],
            "exclusiveTests": Object {
              "matches": false,
              "max": true,
              "min": true,
              "trim": false,
            },
            "spec": Object {
              "abortEarly": true,
              "nullable": false,
              "presence": "optional",
              "recursive": true,
              "strict": false,
              "strip": false,
            },
            "tests": Array [
              [Function],
              [Function],
              [Function],
              [Function],
            ],
            "transforms": Array [
              [Function],
              [Function],
            ],
            "type": "string",
          },
        },
        "spec": Object {
          "abortEarly": true,
          "nullable": false,
          "presence": "optional",
          "recursive": true,
          "strict": false,
          "strip": false,
        },
        "tests": Array [],
        "transforms": Array [
          [Function],
        ],
        "type": "object",
      }
    }
  >
    <Component />
  </Formik>
  <div
    className="loginButton"
  >
    <Link
      className="loginLink"
      to="/login"
    >
      Already have an account? Sign in
    </Link>
  </div>
  <ErrorMessage
    text=""
    type="form"
  />
</div>
`;
